src/Storages/Freeze.cpp

```diff
@@ -0,0 +1,201 @@
+        return false;
+    auto metadata_str = metadata_storage->readFileToString(file_path);
+    ReadBufferFromString buffer(metadata_str);
+    readIntText(version, buffer);
+    if (version != 1)
+    {
+        LOG_ERROR(&Poco::Logger::get("FreezeMetaData"), "Unknown freezed metadata version: {}", version);
+        return false;
+    }
+    DB::assertChar('\n', buffer);
+    readBoolText(is_replicated, buffer);
+    DB::assertChar('\n', buffer);
+    readBoolText(is_remote, buffer);
+    DB::assertChar('\n', buffer);
+    readString(replica_name, buffer);
+    DB::assertChar('\n', buffer);
+    readString(zookeeper_name, buffer);
+    DB::assertChar('\n', buffer);
+    readString(table_shared_id, buffer);
+    DB::assertChar('\n', buffer);
+    return true;
+}
+
+void FreezeMetaData::clean(DiskPtr data_disk, const String & path)
+{
+    auto metadata_storage = data_disk->getMetadataStorage();
+    auto fname = getFileName(path);
+    if (metadata_storage->exists(fname))
+    {
+        auto tx = metadata_storage->createTransaction();
+        tx->unlinkFile(fname);
+        tx->commit();
+    }
+}
+
+String FreezeMetaData::getFileName(const String & path)
+{
+    return fs::path(path) / "frozen_metadata.txt";
+}
+
+BlockIO Unfreezer::unfreeze(const String & backup_name, ContextPtr local_context)
+{
+    LOG_DEBUG(log, "Unfreezing backup {}", backup_name);
+    auto disks_map = local_context->getDisksMap();
+    Disks disks;
+    for (auto & [name, disk]: disks_map)
+    {
+        disks.push_back(disk);
+    }
+    auto backup_path = fs::path(backup_directory_prefix) / escapeForFileName(backup_name);
+    auto store_path = backup_path / "store";
+
+    PartitionCommandsResultInfo result_info;
+
+    for (const auto & disk: disks)
+    {
+        if (!disk->exists(store_path))
+            continue;
+        for (auto prefix_it = disk->iterateDirectory(store_path); prefix_it->isValid(); prefix_it->next())
+        {
+            auto prefix_directory = store_path / prefix_it->name();
+            for (auto table_it = disk->iterateDirectory(prefix_directory); table_it->isValid(); table_it->next())
+            {
+                auto table_directory = prefix_directory / table_it->name();
+                auto current_result_info = unfreezePartitionsFromTableDirectory([] (const String &) { return true; }, backup_name, {disk}, table_directory, local_context);
+                for (auto & command_result : current_result_info)
+                {
+                    command_result.command_type = "SYSTEM UNFREEZE";
+                }
+                result_info.insert(
+                                result_info.end(),
+                                std::make_move_iterator(current_result_info.begin()),
+                                std::make_move_iterator(current_result_info.end()));
+            }
+        }
+        if (disk->exists(backup_path))
+        {
+            disk->removeRecursive(backup_path);
```

Comment:
It's not clear why do we need to "unfreeze" something with complex logic if later we simply recursively remove everything. Comments are required

Reply:
Recursive removal is very dangerous, maybe avoid this type of code?